### DISCUSSION TITLE
有多线程的使用例子么

### DISCUSSION THREAD

**Discussion by 380wmda999** [Tencent/tquic#301] [Q&A (问题和回答)]
我看官方给的demo simple_server.rs 是基于单线程 遍历事件处理的
如果多线程 官方推荐的应该的怎么处理呢 
是创建一个线程池 来处理 还是 ？
有没有这方面的例子

**Comment by glime-ai** [comment]
官方目前没有提供多线程处理的具体示例。不过，您可以参考以下方法来实现多线程处理：

1. **使用线程池**：可以创建一个线程池来处理不同的连接和事件。这样可以提高并发处理能力。

2. **使用异步编程**：利用Rust的异步编程特性（如`tokio`或`async-std`），可以更高效地处理并发任务。

以下是一个简单的示例，展示如何使用`tokio`来实现多线程处理：

```rust
use tokio::net::TcpListener;
use tokio::sync::mpsc;
use std::sync::Arc;
use tquic::Connection;

#[tokio::main]
async fn main() {
    let listener = TcpListener::bind("127.0.0.1:8443").await.unwrap();
    let (tx, mut rx) = mpsc::channel(32);
    let handler = Arc::new(MyHandler { tx });

    loop {
        let (socket, _) = listener.accept().await.unwrap();
        let handler = handler.clone();
        tokio::spawn(async move {
            let conn = Connection::new(socket, handler).await.unwrap();
            conn.run().await.unwrap();
        });
    }
}

struct MyHandler {
    tx: mpsc::Sender<()>,
}

impl tquic::TransportHandler for MyHandler {
    fn on_conn_created(&mut self, conn: &mut Connection) {
        // 处理连接创建事件
    }

    fn on_conn_established(&mut self, conn: &mut Connection) {
        // 处理连接建立事件
    }

    fn on_conn_closed(&mut self, conn: &mut Connection) {
        // 处理连接关闭事件
    }

    fn on_stream_created(&mut self, conn: &mut Connection, stream_id: u64) {
        // 处理流创建事件
    }

    fn on_stream_readable(&mut self, conn: &mut Connection, stream_id: u64) {
        // 处理流可读事件
    }

    fn on_stream_writable(&mut self, conn: &mut Connection, stream_id: u64) {
        // 处理流可写事件
    }
}
```

这个示例使用`tokio`来监听TCP连接，并为每个连接创建一个新的异步任务来处理。您可以根据需要扩展这个示例来处理更多的事件和逻辑。

更多详细信息和示例代码可以参考官方文档和示例：
- [TQUIC Rust 示例代码](https://github.com/tquic-group/tquic-example-rust)
- [TQUIC 文档](https://tquic.net/zh/docs/tutorial/rust)

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
目前没有多线程的示例server。后续版本会提供。

> 如果多线程 官方推荐的应该的怎么处理呢

Server可以创建多个线程，各线程创建独立的Endpoint并运行各自的事件循环

**Reply by lion117** [reply]
libevent 在C++环境下, 使用多线程限制颇多. 是否可以改用libuv来实现client 的多线程?

**Reply by iyangsj** [reply]
可以

**Reply by 380wmda999** [reply]
> Server可以创建多个线程，各线程创建独立的Endpoint并运行各自的事件循环

感觉还是有点抽象 如果一个server创建多个Endpoint 能保证数据的一致性么？ 比如 同一个用户 他的连接id 是固定的  但是 ip和端口是变动的(因为运营性udp有qos quic是根据连接id 动态变动用户的ip和端口 这个也是quic的连接迁移) 我看了下endpoint的pktInfo里面并没有连接id 这样就没法保证同一个用户的数据 是发送到同一个的Endpoint 里面？

**Reply by iyangsj** [reply]
Server多个线程间的报文路由，既可以基于eBPF实现，也可以基于应用层实现（都超出TQUIC协议库范围）
后续版本会提供这两种形式的示例server程序

**Reply by 380wmda999** [reply]
我是想在应用层实现的 根据cpu的个数 动态创建对应的线程 然后每个线程 运行自己的Endpoint处理数据 然后根据quic的连接id  随机到固定的Endpoint实现 但是 看了下文档 好像没发现有什么api可以从外部获取quic的连接id

**Reply by 380wmda999** [reply]
我知道这个确实超出TQUIC协议库范围 但是实际运用中 server肯定是要使用多线程避免因为单个连接阻塞 而影响到其他连接 主要是国内运营商 udp又有qos 同一个人的外网ip和端口会变 不然直接根据外网ip和端口定位到固定的Endpoint实现也没啥问题
我现在用的quic-go实现了一份 quinn也实现了一份 发现 quinn的性能确实比go的吞吐量高一点 现在打算再用tquic实现一份再对比一下

**Reply by iyangsj** [reply]
> 看了下文档 好像没发现有什么api可以从外部获取quic的连接id

可以根据RFC 8999规定，直接从UDP报文内容中提取相应字节。
https://datatracker.ietf.org/doc/html/rfc8999#name-short-header

**Reply by 380wmda999** [reply]
这样确实可以实现 但是解析quic连接id 这应该是协议栈做的事情 最好能提供一个api调用是最好的
